Source/JavaScriptCore/builtins/ModuleLoader.js

```diff
@@ -23,586 +23,642 @@
+    @assert(!!entry.asyncEvaluation);
+    @assert(entry.evaluationError === null);
+
+    entry.evaluationError = { value: error };
+    entry.status = @ModuleStatusEvaluated;
+
+    for (var i = 0, { length } = entry.asyncParentModules; i < length; i++) {
+        var parentEntry = entry.asyncParentModules[i];
+        @asyncModuleExecutionRejected(parentEntry, error);
+    }
+
+    if (entry.topLevelCapability !== null) {
+        @assert(entry.cycleRoot === entry);
+        entry.topLevelCapability.reject(error);
+    }
+}
+
 // APIs to control the module loader.
 
 @visibility=PrivateRecursive
-function provideFetch(key, value)
+function fetchModule(key, parameters, fetcher)
 {
     "use strict";
 
-    var entry = this.ensureRegistered(key);
-
-    if (entry.state > @ModuleFetch)
-        @throwTypeError("Requested module is already fetched.");
-    @fulfillFetch(entry, value);
+    return @ensureImportMapsLoaded(() => {
+        return this.hostFetchAndLoadImportedModule(key, parameters, fetcher).then(entry => {
+            if (entry.errorToRethrow)
+                throw entry.errorToRethrow;
+            return this.fetchDescendantsOfAndLink(entry, fetcher);
+        });
+    });
 }
 
 @visibility=PrivateRecursive
-async function loadModule(key, parameters, fetcher)
+function fetchModuleAndEvaluate(key, parameters, fetcher)
 {
     "use strict";
 
-    var importMap = @importMapStatus();
-    if (importMap)
-        await importMap;
-    var entry = await this.requestSatisfy(this.ensureRegistered(key), parameters, fetcher, new @Set);
-    return entry.key;
+    return this.fetchModule(key, parameters, fetcher).then(() => {
+        var entry = this.getModuleMapEntry(key, this.getModuleType(parameters));
+        @assert(!!entry);
+        return this.moduleEvaluation(entry, fetcher);
+    });
 }
 
 @visibility=PrivateRecursive
-function linkAndEvaluateModule(key, fetcher)
+function evaluateModule(key, fetcher)
 {
     "use strict";
 
-    var entry = this.ensureRegistered(key);
-    this.link(entry, fetcher);
+    var entry = this.moduleMap.@get(key);
+    @assert(!!entry);
     return this.moduleEvaluation(entry, fetcher);
 }
 
 @visibility=PrivateRecursive
-async function loadAndEvaluateModule(moduleName, parameters, fetcher)
+function loadModule(key, sourceCode, fetcher)
+{
+    "use strict";
+
+    var entry = this.ensureModuleMapEntry(key, @ModuleTypeJavaScript);
+    entry.fetchAndParsePromise ??= this.parseModule(key, sourceCode).then(moduleRecord => this.hostLoadImportedModule(entry, moduleRecord), error => {
+        entry.errorToRethrow = error;
+        return entry;
+    });
+
+    return @ensureImportMapsLoaded(() => {
+        return entry.fetchAndParsePromise
+            .then(entry => {
+                if (entry.errorToRethrow)
+                    throw entry.errorToRethrow;
+                return entry;
+            })
+            .then(() => this.fetchDescendantsOfAndLink(entry, fetcher));
```

Comment:
Maybe use `.then((entry) => this.fetchDescendantsOfAndLink(entry, fetcher));` is better?

Reply:
Yes, totally, it's slightly faster to avoid scope lookup and prefer local variables. Fixed.